### ISSUE TITLE
[MNG-5649] Use Commons Lang's Validate to intercept invalid input

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6668]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** opened **[MNG-5649](https://issues.apache.org/jira/browse/MNG-5649?redirect=false)** and commented

In several spots of Maven Core IAE is thrown where an argument is null. This should be turned into `NullPointerException` since JDK adheres to is and the [description](http://docs.oracle.com/javase/6/docs/api/java/lang/NullPointerException.html) of this exception indicates that and Effective Java does that too.

I possible fix version could next minor: 3.3. Is no one is opposed it could even be 3.2.2.

---

**Affects:** 3.2.1

**Issue Links:**
- [MRESOLVER-3](https://issues.apache.org/jira/browse/MRESOLVER-3) Update dependencies

- [MNG-6829](https://issues.apache.org/jira/browse/MNG-6829) Remove commons-lang3 dependency

1 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

```
$ grep -nrF "throw new IllegalArgumentException" .
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultArtifactDescriptorReader.java:168:
        throw new IllegalArgumentException( "remote repository manager has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultArtifactDescriptorReader.java:178:
        throw new IllegalArgumentException( "version resolver has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultArtifactDescriptorReader.java:189:
        throw new IllegalArgumentException( "version range resolver has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultArtifactDescriptorReader.java:199:
        throw new IllegalArgumentException( "artifact resolver has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultArtifactDescriptorReader.java:209:
        throw new IllegalArgumentException( "repository event dispatcher has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultArtifactDescriptorReader.java:219:
        throw new IllegalArgumentException( "model builder has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java:130:
    throw new IllegalArgumentException( "metadata resolver has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java:140:
    throw new IllegalArgumentException( "sync context factory has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java:150:
    throw new IllegalArgumentException( "repository event dispatcher has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionResolver.java:142:            t
hrow new IllegalArgumentException( "metadata resolver has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionResolver.java:152:            t
hrow new IllegalArgumentException( "sync context factory has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionResolver.java:162:            t
hrow new IllegalArgumentException( "repository event dispatcher has not been specified" );
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/RelocatedArtifact.java:47:            throw n
ew IllegalArgumentException( "no artifact specified" );
./maven-artifact/src/main/java/org/apache/maven/artifact/ArtifactUtils.java:55:            throw new IllegalArgumentExce
ption( "version: null" );
./maven-core/src/main/java/org/apache/maven/artifact/repository/LegacyLocalRepositoryManager.java:93:            throw n
ew IllegalArgumentException( "local repository delegate missing" );
./maven-core/src/main/java/org/apache/maven/artifact/repository/metadata/io/DefaultMetadataReader.java:50:            th
row new IllegalArgumentException( "input file missing" );
./maven-core/src/main/java/org/apache/maven/artifact/repository/metadata/io/DefaultMetadataReader.java:63:            th
row new IllegalArgumentException( "input reader missing" );
./maven-core/src/main/java/org/apache/maven/artifact/repository/metadata/io/DefaultMetadataReader.java:86:            th
row new IllegalArgumentException( "input stream missing" );
./maven-core/src/main/java/org/apache/maven/classrealm/DefaultClassRealmManager.java:312:            throw new IllegalAr
gumentException( "model missing" );
./maven-core/src/main/java/org/apache/maven/classrealm/DefaultClassRealmManager.java:329:            throw new IllegalAr
gumentException( "extension plugin missing" );
./maven-core/src/main/java/org/apache/maven/classrealm/DefaultClassRealmManager.java:345:            throw new IllegalAr
gumentException( "plugin missing" );
./maven-core/src/main/java/org/apache/maven/configuration/DefaultBeanConfigurationRequest.java:125:            throw new
 IllegalArgumentException( "group id for plugin has not been specified" );
./maven-core/src/main/java/org/apache/maven/configuration/DefaultBeanConfigurationRequest.java:129:            throw new
 IllegalArgumentException( "artifact id for plugin has not been specified" );
./maven-core/src/main/java/org/apache/maven/configuration/internal/DefaultBeanConfigurator.java:58:            throw new
 IllegalArgumentException( "bean configuration request not specified" );
./maven-core/src/main/java/org/apache/maven/configuration/internal/DefaultBeanConfigurator.java:63:            throw new
 IllegalArgumentException( "bean to be configured not specified" );
./maven-core/src/main/java/org/apache/maven/DefaultProjectDependencyGraph.java:65:            throw new IllegalArgumentE
xception( "project missing" );
./maven-core/src/main/java/org/apache/maven/DefaultProjectDependencyGraph.java:90:            throw new IllegalArgumentE
xception( "project missing" );
./maven-core/src/main/java/org/apache/maven/execution/BuildSummary.java:52:            throw new IllegalArgumentExceptio
n( "project missing" );
./maven-core/src/main/java/org/apache/maven/execution/DefaultMavenExecutionRequest.java:706:            throw new Illega
lArgumentException( "proxy missing" );
./maven-core/src/main/java/org/apache/maven/execution/DefaultMavenExecutionRequest.java:749:            throw new Illega
lArgumentException( "server missing" );
./maven-core/src/main/java/org/apache/maven/execution/DefaultMavenExecutionRequest.java:792:            throw new Illega
lArgumentException( "mirror missing" );
./maven-core/src/main/java/org/apache/maven/execution/DefaultMavenExecutionRequest.java:1008:            throw new Illeg
alArgumentException( "profile missing" );
./maven-core/src/main/java/org/apache/maven/execution/ReactorManager.java:103:            throw new IllegalArgumentExcep
tion( "Invalid failure behavior (must be one of: \'" + FAIL_FAST + "\', \'"
./maven-core/src/main/java/org/apache/maven/FilteredProjectDependencyGraph.java:57:            throw new IllegalArgument
Exception( "project dependency graph missing" );
./maven-core/src/main/java/org/apache/maven/lifecycle/internal/builder/BuilderCommon.java:173:            throw new Ille
galArgumentException(
./maven-core/src/main/java/org/apache/maven/plugin/DefaultPluginRealmCache.java:167:            throw new IllegalArgumen
tException();
./maven-core/src/main/java/org/apache/maven/plugin/internal/DefaultMavenPluginManager.java:344:            throw new Ill
egalArgumentException( "incomplete plugin descriptor, plugin missing" );
./maven-core/src/main/java/org/apache/maven/plugin/internal/DefaultMavenPluginManager.java:351:            throw new Ill
egalArgumentException( "incomplete plugin descriptor, plugin artifact missing" );
./maven-core/src/main/java/org/apache/maven/project/DefaultModelBuildingListener.java:58:            throw new IllegalAr
gumentException( "project missing" );
./maven-core/src/main/java/org/apache/maven/project/DefaultModelBuildingListener.java:64:            throw new IllegalAr
gumentException( "project building helper missing" );
./maven-core/src/main/java/org/apache/maven/project/DefaultModelBuildingListener.java:70:            throw new IllegalAr
gumentException( "project building request missing" );
./maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuildingRequest.java:325:            throw new Illegal
ArgumentException( "repository merge mode not specified" );
./maven-core/src/main/java/org/apache/maven/project/MavenProject.java:253:            throw new IllegalArgumentException
( "mavenTools: null" );
./maven-core/src/main/java/org/apache/maven/repository/ArtifactTransferEvent.java:151:                throw new IllegalA
rgumentException( "Illegal request type: " + requestType );
./maven-core/src/main/java/org/apache/maven/repository/ArtifactTransferEvent.java:184:                throw new IllegalA
rgumentException( "Illegal event type: " + eventType );
./maven-core/src/main/java/org/apache/maven/rtinfo/internal/DefaultRuntimeInformation.java:109:            throw new Ill
egalArgumentException( "Version range must not be null" );
./maven-core/src/main/java/org/apache/maven/rtinfo/internal/DefaultRuntimeInformation.java:113:            throw new Ill
egalArgumentException( "Version range must not be empty" );
./maven-core/src/main/java/org/apache/maven/rtinfo/internal/DefaultRuntimeInformation.java:123:            throw new Ill
egalArgumentException( e.getMessage(), e );
./maven-core/src/test/java/org/apache/maven/configuration/DefaultBeanConfiguratorTest.java:66:            throw new Ille
galArgumentException( e );
./maven-core/src/test/java/org/apache/maven/configuration/DefaultBeanConfiguratorTest.java:70:            throw new Ille
galArgumentException( e );
./maven-core/src/test/java/org/apache/maven/project/harness/PomTestWrapper.java:48:            throw new IllegalArgument
Exception( "mavenProject: null" );
./maven-core/src/test/java/org/apache/maven/project/harness/PomTestWrapper.java:59:            throw new IllegalArgument
Exception( "mavenProject: null" );
./maven-core/src/test/java/org/apache/maven/project/harness/PomTestWrapper.java:86:            throw new IllegalArgument
Exception( "uri: null or empty" );
./maven-embedder/src/main/java/org/apache/maven/cli/event/ExecutionEventLogger.java:63:            throw new IllegalArgu
mentException( "logger missing" );
./maven-model/src/main/java/org/apache/maven/model/merge/ModelMerger.java:106:            throw new IllegalArgumentExcep
tion( "target missing" );
./maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:466:                throw n
ew IllegalArgumentException( "neither model source nor input file are specified" );
./maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:895:            throw new I
llegalArgumentException( "no model resolver provided, cannot resolve parent POM "
./maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:1068:                    th
row new IllegalArgumentException( "no model resolver provided, cannot resolve import POM "
./maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuildingResult.java:81:            throw
 new IllegalArgumentException( "no model identifier specified" );
./maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuildingResult.java:103:            thro
w new IllegalArgumentException( "no model identifier specified" );
./maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuildingResult.java:120:            thro
w new IllegalArgumentException( "no model identifier specified" );
./maven-model-builder/src/main/java/org/apache/maven/model/building/FileModelSource.java:47:            throw new Illega
lArgumentException( "no POM file specified" );
./maven-model-builder/src/main/java/org/apache/maven/model/building/UrlModelSource.java:46:            throw new Illegal
ArgumentException( "no POM URL specified" );
./maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java:53:            throw new IllegalAr
gumentException( "input file missing" );
./maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java:68:            throw new IllegalAr
gumentException( "input reader missing" );
./maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java:86:            throw new IllegalAr
gumentException( "input stream missing" );
./maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelWriter.java:50:            throw new IllegalAr
gumentException( "output file missing" );
./maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelWriter.java:55:            throw new IllegalAr
gumentException( "model missing" );
./maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelWriter.java:68:            throw new IllegalAr
gumentException( "output writer missing" );
./maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelWriter.java:73:            throw new IllegalAr
gumentException( "model missing" );
./maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelWriter.java:92:            throw new IllegalAr
gumentException( "output stream missing" );
./maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelWriter.java:97:            throw new IllegalAr
gumentException( "model missing" );
./maven-model-builder/src/test/java/org/apache/maven/model/profile/activation/AbstractProfileActivatorTest.java:50:
       throw new IllegalArgumentException( "class of profile activator to test is not specified" );
./maven-settings/src/main/mdo/settings.mdo:73:            throw new IllegalArgumentException( "sourceLevel must be one o
f: {" + USER_LEVEL + "," + GLOBAL_LEVEL + "}" );
./maven-settings/target/generated-sources/modello/org/apache/maven/settings/TrackableBase.java:62:            throw new
IllegalArgumentException( "sourceLevel must be one of: {" + USER_LEVEL + "," + GLOBAL_LEVEL + "}" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/building/FileSettingsSource.java:47:            throw n
ew IllegalArgumentException( "no settings file specified" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/building/UrlSettingsSource.java:46:            throw ne
w IllegalArgumentException( "no settings URL specified" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/io/DefaultSettingsReader.java:50:            throw new
IllegalArgumentException( "input file missing" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/io/DefaultSettingsReader.java:63:            throw new
IllegalArgumentException( "input reader missing" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/io/DefaultSettingsReader.java:86:            throw new
IllegalArgumentException( "input stream missing" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/io/DefaultSettingsWriter.java:50:            throw new
IllegalArgumentException( "output file missing" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/io/DefaultSettingsWriter.java:55:            throw new
IllegalArgumentException( "settings missing" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/io/DefaultSettingsWriter.java:68:            throw new
IllegalArgumentException( "output writer missing" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/io/DefaultSettingsWriter.java:73:            throw new
IllegalArgumentException( "settings missing" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/io/DefaultSettingsWriter.java:92:            throw new
IllegalArgumentException( "output stream missing" );
./maven-settings-builder/src/main/java/org/apache/maven/settings/io/DefaultSettingsWriter.java:97:            throw new
IllegalArgumentException( "settings missing" );
```

Almost in every case listed above. Guava's `Preconditions` would solve the issue. I use Commons Lang's `Validate` in my projects.  When we switch to Java 7, we could simply use [this](http://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#requireNonNull(T)).

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Jason, several months have passed. Are still interested in Guava's Preconditions? Otherwise I'd go over to fix those issues.

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

Why not use it then ? Guava is a bit of a dead-end project come jdk8

**Comment by jira-importer** [comment]
**[Mark Struberg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=struberg)** commented

-1 for using Guava all over. I've used in in the past and it turned out to introduce way more backward incompatibility headache than commons. And IllegalArgumentException is btw much more explicit than a NullPointerException.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Interesting to say is that this ticket has been open for a long time and no one except me and Jason did ever comment on it. Removing Guava would require to another ticket first. I am good to do that.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Using IAE is a bad thing. High discouraged. Just read Effective Java. Always keep in mind that unchecked exceptions indicate programming errors and a null variable will always lead a NPE.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

For the typical check on null arguments i would suggest to use the default JDK 7 solution, and yes this will throw an NPE, cause it is an NPE with an appropriate message exactly shows what is wrong...

```java
this.whatever = Objects.requireNoneNull(...);
```

which is also available in JDK8/JDK9...

And for the others i would vote for Apache Commons ?...

I have checked the code at the moment we have the following references to `com.google.*`

```
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultModelResolver.java
./maven-aether-provider/src/main/java/org/apache/maven/repository/internal/MavenAetherModule.java
./maven-core/src/main/java/org/apache/maven/DefaultMaven.java
./maven-core/src/main/java/org/apache/maven/SessionScoped.java
./maven-core/src/main/java/org/apache/maven/execution/DefaultMavenExecutionRequest.java
./maven-core/src/main/java/org/apache/maven/execution/scope/MojoExecutionScoped.java
./maven-core/src/main/java/org/apache/maven/execution/scope/internal/MojoExecutionScope.java
./maven-core/src/main/java/org/apache/maven/execution/scope/internal/MojoExecutionScopeModule.java
./maven-core/src/main/java/org/apache/maven/extension/internal/CoreExports.java
./maven-core/src/main/java/org/apache/maven/extension/internal/CoreExtensionEntry.java
./maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java
./maven-core/src/main/java/org/apache/maven/lifecycle/DefaultLifecycles.java
./maven-core/src/main/java/org/apache/maven/lifecycle/internal/DefaultLifecycleExecutionPlanCalculator.java
./maven-core/src/main/java/org/apache/maven/project/ProjectModelResolver.java
./maven-core/src/main/java/org/apache/maven/session/scope/internal/SessionScope.java
./maven-core/src/main/java/org/apache/maven/session/scope/internal/SessionScopeModule.java
./maven-core/src/test/java/org/apache/maven/execution/scope/internal/MojoExecutionScopeTest.java
./maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java
./maven-model-builder/src/main/java/org/apache/maven/model/building/Result.java
```

In other words most of the references to Google Guava shouldn't be a big issue to remove / replace with others...

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I went for `Objects.requireNotNull` first but if I will integrate Commons Lang, therefore I will use `Verify` only to make the code consistent.

Regarding to the remaining Guava code -- yes it is doable but I hardly doubt without external libraries (Commons Collections). Otherwise one would reinvent the wheel again.

As there seems to be a tend for Commons Lang and no objections, I would go for Commons Lang. For the removal of Guava, we shall spawn another ticket anyway.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

I don't say without external libraries...where i would prefer (Commons.*) instead of Google Guava...Yes for an other ticket for Guava removing...

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Agreed!

**Comment by jira-importer** [comment]
**[Uwe Barthel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=barthel)** commented

+1 for Apache Commons. It's also a kind of eat-your-own-dog-food thing. ;-)

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Do you need any assistance here?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I have created an appropriate branch where I will push commit. After they are stabilized, I will squash and merge back.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Hi Michael, i have committed on your branch can you please review the changes i have made. Thanks.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

That wasn't a smart move because you have replicated all code changes I have committed already locally. The status `In Progress` has been set on purpose. Please reset the branch to its previous state and force push it. I will push shortly.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Done.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Changes squashed in one commit and force pushed to the branch. Have a look. If you have no objections, I'd merge into master, push and drop the branch.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Go for it.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [618e62dd3315b4cb5b0f7dcdd37fc787c44b2ade](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commit;h=618e62dd3315b4cb5b0f7dcdd37fc787c44b2ade).

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

You can now go ahead and create another issue for the removal of Guava. Commons Collections should suffice.
